modules/Walking_module/src/WalkingQPInverseKinematics.cpp

```diff
@@ -232,26 +455,255 @@ bool WalkingQPIK::setDesiredJointPosition(const iDynTree::VectorDynSize& regular
 {
-    m_comVelocity = comVelocity;
+    m_desiredComVelocity = comVelocity;
 }
 
 void WalkingQPIK::setDesiredCoMPosition(const iDynTree::Position& desiredComPosition)
 {
     m_desiredComPosition = desiredComPosition;
 }
 
+void WalkingQPIK::evaluateHessianMatrix()
+{
+    // in that case the hessian matrix is related only to neck orientation and to
+    // the joint angle
+    auto hessianDense(iDynTree::toEigen(m_hessianDense));
+
+    hessianDense = iDynTree::toEigen(m_neckJacobian).transpose() *
+        m_neckWeight *
+        iDynTree::toEigen(m_neckJacobian);
+
+    if(!m_enableHandRetargeting)
+        hessianDense += iDynTree::toEigen(m_jointRegularizationHessian);
+    else
+    {
+        // think about the possibility to project in the null space the joint regularization
+        hessianDense += iDynTree::toEigen(m_jointRegularizationHessian)
+            + iDynTree::toEigen(m_leftHandJacobian).transpose()
+            * iDynTree::toEigen(m_handWeightSmoother->getPos()).asDiagonal()
+            * iDynTree::toEigen(m_leftHandJacobian)
+            + iDynTree::toEigen(m_rightHandJacobian).transpose()
+            * iDynTree::toEigen(m_handWeightSmoother->getPos()).asDiagonal()
+            * iDynTree::toEigen(m_rightHandJacobian);
+    }
+
+    if(!m_useCoMAsConstraint)
+    {
+        hessianDense += iDynTree::toEigen(m_comJacobian).transpose() *
+            iDynTree::toEigen(m_comWeight).asDiagonal() * iDynTree::toEigen(m_comJacobian);
+    }
+}
+
+void WalkingQPIK::evaluateGradientVector()
+{
+    auto gradient(iDynTree::toEigen(m_gradient));
+
+    auto neckJacobian(iDynTree::toEigen(m_neckJacobian));
+
+    auto jointRegularizationGradient(iDynTree::toEigen(m_jointRegularizationGradient));
+    auto jointRegularizationGains(iDynTree::toEigen(m_jointRegularizationGains));
+    auto jointPosition(iDynTree::toEigen(m_jointPosition));
+    auto regularizationTerm(iDynTree::toEigen(m_regularizationTerm));
+
+    auto comJacobian(iDynTree::toEigen(m_comJacobian));
+    auto comWeight(iDynTree::toEigen(m_comWeight));
+    auto comPosition(iDynTree::toEigen(m_comPosition));
+    auto desiredComPosition(iDynTree::toEigen(m_desiredComPosition));
+    auto desiredComVelocity(iDynTree::toEigen(m_desiredComVelocity));
+
+    // Neck orientation
+    iDynTree::Matrix3x3 errorNeckAttitude = iDynTreeHelper::Rotation::skewSymmetric(m_neckOrientation * m_desiredNeckOrientation.inverse());
+    gradient = -neckJacobian.transpose() * m_neckWeight * (-m_kNeck
+                                                           * iDynTree::unskew(iDynTree::toEigen(errorNeckAttitude)));
+
+    // if the hand retargeting is not enable the regularization term should be used as usual
+    if(!m_enableHandRetargeting)
+       gradient += -jointRegularizationGradient * jointRegularizationGains.asDiagonal()
+           * (regularizationTerm - jointPosition);
+    else
+    {
+        //  left hand
+        Eigen::VectorXd leftHandCorrection(6);
```

Comment:
Why not using a fixed size vector here?

Reply:
even better: `leftHandCorrection` may become an `iDynTree::Twist` object and it can belong to the class attributes.